DemoInfo/DemoParser.cs

```diff
@@ -506,9 +527,142 @@ public void ParseHeader()
 
 			Header = header;
 
-
 			if (HeaderParsed != null)
 				HeaderParsed(this, new HeaderParsedEventArgs(Header));
+
+				if (header.PlaybackTime == 0)  //header is corrupted
+					FixHeader();
+		}
+
+		void FixHeader()
+		{
+			ExpressParseToEnd();
+			Header.PlaybackTicks = IngameTick;
+			Header.PlaybackFrames = CurrentTick - (TickAtStart + 1);
+			Header.PlaybackTime = TickInterval * IngameTick;
+			
+			CurrentTick = 0;
+
+			//the ol' switcheroo
+			//It's probably safest and easiest to
+			//swap out for a fresh bitstream.
+			Input.Seek(0, System.IO.SeekOrigin.Begin);
```

Comment:
Perhaps I'm missing something obvious but it's still not clear to me why we actually *need* to restart here. It's been a while since I've last touched the internals of this but as far as I remember, the parser itself doesn't actually rely on the correct header data being there, does it?

Sure, we hand out the bad data from the demo header but I'd say most applications can handle getting this info only later during parsing (when the ServerInfo message is parsed). Those who can't still have the option of restarting the parser from the outside and it's actually even easier for them as they don't have to worry about internal object state.